internal/datastore/proxy/relationshipintegrity.go

```diff
@@ -0,0 +1,482 @@
+	// Ensure the datastore supports integrity.
+	unwrapped := ds
+	if wds, ok := unwrapped.(datastore.UnwrappableDatastore); ok {
+		unwrapped = wds.Unwrap()
+	}
+
+	id, ok := unwrapped.(datastore.IntegrityDatastore)
+	if !ok {
+		return nil, fmt.Errorf("datastore of kind %T does not implement relationship integrity", ds)
+	}
+
+	if !id.SupportsIntegrity() {
+		return nil, spiceerrors.MustBugf("datastore does not support relationship integrity")
+	}
+
+	if len(currentKey.Bytes) == 0 {
+		return nil, fmt.Errorf("contents of the current key file cannot be empty")
+	}
+
+	if len(currentKey.ID) == 0 {
+		return nil, fmt.Errorf("current key ID cannot be empty")
+	}
+
+	hasher, err := hasherForKey(currentKey.Bytes)
+	if err != nil {
+		return nil, err
+	}
+
+	currentKeyHMAC := &hmacConfig{
+		keyID:     currentKey.ID,
+		expiredAt: currentKey.ExpiredAt,
+		hasher:    hasher,
+		lock:      sync.Mutex{},
+	}
+
+	if currentKey.ExpiredAt != nil {
+		return nil, spiceerrors.MustBugf("current key cannot have an expiration")
+	}
+
+	keysByID := make(map[string]*hmacConfig, len(expiredKeys)+1)
+	keysByID[currentKey.ID] = currentKeyHMAC
+
+	expiredKeyIDs := make([]string, 0, len(expiredKeys))
+	for _, key := range expiredKeys {
+		if len(key.Bytes) == 0 {
+			return nil, fmt.Errorf("expired key cannot be empty")
+		}
+
+		if len(key.ID) == 0 {
+			return nil, fmt.Errorf("expired key ID cannot be empty")
+		}
+
+		if key.ExpiredAt == nil {
+			return nil, fmt.Errorf("expired key missing expiration time")
+		}
+
+		if _, ok := keysByID[key.ID]; ok {
+			return nil, fmt.Errorf("found duplicate key ID: %s", key.ID)
+		}
+
+		hasher, err := hasherForKey(key.Bytes)
+		if err != nil {
+			return nil, err
+		}
+
+		keysByID[key.ID] = &hmacConfig{
+			keyID:     key.ID,
+			expiredAt: key.ExpiredAt,
+			hasher:    hasher,
+		}
+
+		expiredKeyIDs = append(expiredKeyIDs, key.ID)
+	}
+
+	log.Debug().
+		Str("current_key_id", currentKey.ID).
+		Strs("expired_key_ids", expiredKeyIDs).
+		Msgf("created relationship integrity proxy")
+
+	return &relationshipIntegrityProxy{
+		ds:         ds,
+		primaryKey: currentKeyHMAC,
+		keysByID:   keysByID,
+	}, nil
+}
+
+func hasherForKey(key []byte) (hash.Hash, error) {
+	return hmac.New(sha256.New, key), nil
+}
+
+type relationshipIntegrityProxy struct {
+	ds         datastore.Datastore
+	primaryKey *hmacConfig
+	keysByID   map[string]*hmacConfig
+}
+
+func (r *relationshipIntegrityProxy) lookupKey(keyID string) (*hmacConfig, error) {
+	key, ok := r.keysByID[keyID]
+	if !ok {
+		return nil, fmt.Errorf("key not found: %s", keyID)
+	}
+
+	return key, nil
+}
+
+// computeRelationshipHash computes the HMAC hash of a relationship tuple.
+func computeRelationshipHash(tpl *corev1.RelationTuple, key *hmacConfig) ([]byte, error) {
+	bytes, err := tuple.CanonicalBytes(tpl)
+	if err != nil {
+		return nil, err
+	}
+
+	key.lock.Lock()
```

Comment:
this seems like it would be a bottleneck for large reads. what about a pool of hashers instead? either a `sync.Pool` or even just a `chan hasher`

Reply:
I had a pool of them at first, but removed it because I didn't think we needed to lock it. I've added the pool back in